content/microcontrollers/bluepill.md

```diff
@@ -29,3 +29,44 @@ Programs are loaded onto the Bluepill with a separate hardware programmer such a
 - Plug your STLink v2 programmer into your computer's USB port.
 - Plug your Bluepill into the STLink v2 programmer using the Bluepill `SWIO`, `SWCLK`, `3V3` and `GND` pins.
 - Build and flash your TinyGo program using `tinygo flash -target=bluepill`
+
+## OpenOCD Possible Modification
+
+If you run the flasher, and you are getting the error **Warn : UNEXPECTED idcode: 0x2ba01477** you can mitigate the problem by editing the stm32f1x.cfg file within openOCD
+```sh
+> tinygo flash -target=bluepill -ocd-output main.go
+Open On-Chip Debugger 0.10.0
+Licensed under GNU GPL v2
+For bug reports, read
+        http://openocd.org/doc/doxygen/bugs.html
+Info : auto-selecting first available session transport "hla_swd". To override use 'transport select <transport>'.
+Info : The selected transport took over low-level target control. The results might differ compared to plain JTAG/SWD
+adapter speed: 1000 kHz
+adapter_nsrst_delay: 100
+none separate
+Info : Unable to match requested speed 1000 kHz, using 950 kHz
+Info : Unable to match requested speed 1000 kHz, using 950 kHz
+Info : clock speed 950 kHz
+Info : STLINK v2 JTAG v29 API v2 SWIM v7 VID 0x0483 PID 0x3748
+Info : using stlink api v2
+Info : Target voltage: 3.160026
+Warn : UNEXPECTED idcode: 0x2ba01477
+Error: expected 1 of 1: 0x1ba01477
+in procedure 'program' 
+in procedure 'init' called at file "embedded:startup.tcl", line 506
+in procedure 'ocd_bouncer'
+** OpenOCD init failed **
+shutdown command invoked
+
+```
+
+
+```sh
+> sudo vim /usr/share/openocd/scripts/target/stm32f1x.cfg
+```
+
+Edit the line **set _CPUTAPID 0x1ba01477** and change the value to **set _CPUTAPID 0x2ba01477**.  This is usually a known problem with the chip on blue pills, since there are a lot of different chips out there.  This way, you can still write to those microprocessors.  
```

Comment:
If we could override the `CPUTAPID` value passed to OpenOCD to set it to `0` then OpenOCD will ignore the ID when flashing. That would be a better solution for users, no doubt.

@aykevl what do you think?

Reply:
I'll definitely test it out if you guys make the argument change.

Reply:
... I'll see if I can make the change in my fork, and recompile tinygo and test it out today.

Reply:
Do you have any examples of the openocd command that (i think) flashes from /tmp/tingoNNNNNNNNNN/main.hex?  

I can see that the /usr/share/openocd/scripts/target/stm32f1x.cfg checks the variable CPUTAPID first, though when I change my /etc/environment to reflect CPUTAPID="the correct id for my st-link" ... /bin/tclsh doesn't seem to take ANY environment variables from my system.  So unless there is a way to set the var before executing openocd, I'm not sure that there can be a solution unless one just edits the /usr/share/openocd/scripts/target/stm32f1x.cfg manually.

This seems counter intuitive that the writer of openocd would check the variable in tclsh, and one couldn't pass variables into openocd before running, so I might be missing something here.

Reply:
```
STANLEY ⛰  >>> echo $CPUTAPID
0x2ba01477
STANLEY ⛰  >>> /bin/tclsh
% echo $CPUTAPID
can't read "CPUTAPID": no such variable
% set CPUTAPID 0x2ba01477
0x2ba01477
% echo $CPUTAPID
0x2ba01477
%
```

Reply:
..and yeah if there is a way to just add a flag to tinygo, like -openocd-cputapid=0x2ba01477
that would be amazing.

Reply:
How do you feel about setting **all** the CPUTAPID's to zero?  Would that be a workable solution?  Might that break something?  I'm running on an inexpensive st-link v2, and would need to buy a jtag to see if it break it for all the people that use jtags.  I have no idea for sure, but I'm willing to give it a shot.  I'll order one on Amazon RN.

I could elevate user privs (to root [at least on linux]) on regex error match, and then the problem would only need to be fixed one time for all the boards.  Setting _CPUTAPID to zero worked perfectly btw.  Thanks for the suggestion.  I was thinking about global variables, but that is a way better solution.

```
STANLEY ⛰  >>> cd /usr/share/openocd/
STANLEY ⛰  >>> grep -r './' -e '_CPUTAPID 0x'
./scripts/target/stellaris.cfg:   set _CPUTAPID 0x0ba00477
./scripts/target/samsung_s3c2410.cfg:   set _CPUTAPID 0xffffffff
./scripts/target/ti_calypso.cfg:        set _CPUTAPID 0x3100e02f
./scripts/target/nuc910.cfg:   set _CPUTAPID 0x07926f0f
./scripts/target/c100.cfg:   set _CPUTAPID 0x27b3645b
./scripts/target/1986ве1т.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/1986ве1т.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/mdr32f9q2i.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/mdr32f9q2i.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/at91sam9.cfg:  set _CPUTAPID 0x0792603f
./scripts/target/lpc3131.cfg:   set _CPUTAPID 0x07926f0f
./scripts/target/ixp42x.cfg:   set _CPUTAPID 0x19274013
./scripts/target/tmpa910.cfg:   set _CPUTAPID 0x07926031
./scripts/target/imx25.cfg:   set _CPUTAPID 0x07926041
./scripts/target/quark_x10xx.cfg:   set _CPUTAPID 0x18289013
./scripts/target/dsp56321.cfg:   set _CPUTAPID 0x1181501d
./scripts/target/mc13224v.cfg:set _CPUTAPID 0x1f1f001d
./scripts/target/str730.cfg:   set _CPUTAPID 0x3f0f0f0f
./scripts/target/str710.cfg:   set _CPUTAPID 0x3f0f0f0f
./scripts/target/stm32f0x.cfg:   set _CPUTAPID 0x0bb11477
./scripts/target/samsung_s3c6410.cfg:   set _CPUTAPID 0x07b76f0f
./scripts/target/stm32w108xx.cfg:      set _CPUTAPID 0x3ba00477
./scripts/target/stm32w108xx.cfg:      set _CPUTAPID 0x1ba01477
./scripts/target/epc9301.cfg:   set _CPUTAPID 0xffffffff
./scripts/target/imx35.cfg:   set _CPUTAPID 0x07b3601d
./scripts/target/stm32f3x.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/stm32f3x.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/exynos5250.cfg:        set _CPUTAPID 0x4ba00477
./scripts/target/dsp568013.cfg:   set _CPUTAPID 0x01f2401d
./scripts/target/smp8634.cfg:   set _CPUTAPID 0x08630001
./scripts/target/quark_d20xx.cfg:   set _CPUTAPID 0x38289013
./scripts/target/at91sam7x512.cfg:   set _CPUTAPID 0x3f0f0f0f
./scripts/target/cs351x.cfg:   set _CPUTAPID 0x00526fa1
./scripts/target/at91sam3XXX.cfg:   set _CPUTAPID 0x4ba00477
./scripts/target/at91sam4XXX.cfg:   set _CPUTAPID 0x4ba00477
./scripts/target/nrf52.cfg:     set _CPUTAPID 0x2ba01477
./scripts/target/at91r40008.cfg:   set _CPUTAPID 0x1f0f0f0f
./scripts/target/adsp-sc58x.cfg:        set _CPUTAPID 0x3BA02477
./scripts/target/aduc702x.cfg:   set _CPUTAPID 0x3f0f0f0f
./scripts/target/lpc1xxx.cfg:           set _CPUTAPID 0x0bb11477
./scripts/target/lpc1xxx.cfg:                   set _CPUTAPID 0x4ba00477
./scripts/target/lpc1xxx.cfg:                   set _CPUTAPID 0x2ba01477
./scripts/target/pxa255.cfg:   set _CPUTAPID 0x69264013
./scripts/target/fm3.cfg:       set _CPUTAPID 0x4ba00477
./scripts/target/avr32.cfg:set _CPUTAPID 0x21e8203f
./scripts/target/at91sam7a2.cfg:   set _CPUTAPID 0x1f0f0f0f
./scripts/target/nrf51.cfg:   set _CPUTAPID 0x0bb11477
./scripts/target/gp326xxxa.cfg:   set _CPUTAPID 0x4f1f0f0f
./scripts/target/alphascale_asm9260t.cfg:       set _CPUTAPID 0x079264F3
./scripts/target/aducm360.cfg:   set _CPUTAPID 0x2ba01477
./scripts/target/pic32mx.cfg:   set _CPUTAPID 0x30938053
./scripts/target/dragonite.cfg:   set _CPUTAPID 0x121003d3
./scripts/target/imx21.cfg:   set _CPUTAPID 0x0792611f
./scripts/target/klx.cfg:    set _CPUTAPID 0x0bc11477
./scripts/target/atheros_ar2315.cfg:    set _CPUTAPID 0x00000001
./scripts/target/efm32.cfg:   set _CPUTAPID 0x2ba01477
./scripts/target/samsung_s3c2450.cfg:  set _CPUTAPID 0x07926f0f
./scripts/target/stm32l1.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/stm32l1.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/omap5912.cfg:   set _CPUTAPID 0x0692602f
./scripts/target/at91samdXX.cfg:   set _CPUTAPID 0x4ba00477
./scripts/target/sharp_lh79532.cfg:   set _CPUTAPID 0x00002061
./scripts/target/pxa270.cfg:   set _CPUTAPID 0x49265013
./scripts/target/str750.cfg:   set _CPUTAPID 0x4f1f0041
./scripts/target/tmpa900.cfg:   set _CPUTAPID 0x07926031
./scripts/target/psoc4.cfg:   set _CPUTAPID 0x0bb11477
./scripts/target/hilscher_netx50.cfg:   set _CPUTAPID 0x25966021
./scripts/target/ks869x.cfg:   set  _CPUTAPID 0x00922f0f
./scripts/target/kex.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/kex.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/sim3x.cfg:   set _CPUTAPID 0x4ba00477
./scripts/target/imx31.cfg:   set _CPUTAPID 0x07b3601d
./scripts/target/faux.cfg:   set _CPUTAPID 0x00000000
./scripts/target/str912.cfg:   set _CPUTAPID 0x25966041
./scripts/target/stm32f4x.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/stm32f4x.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/at91sam7se512.cfg:   set _CPUTAPID 0xffffffff
./scripts/target/is5114.cfg:   set _CPUTAPID 0xffffffff
./scripts/target/atmega128.cfg:   set _CPUTAPID 0x8970203F
./scripts/target/at91sam3nXX.cfg:       set _CPUTAPID 0x4ba00477
./scripts/target/u8500.cfg:     set _CPUTAPID 0x4ba00477
./scripts/target/k1921vk01t.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/k1921vk01t.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/atsamv.cfg:   set _CPUTAPID 0x0bd11477
./scripts/target/atheros_ar2313.cfg:    set _CPUTAPID 0x00000001
./scripts/target/armada370.cfg:   set _CPUTAPID 0x4ba00477
./scripts/target/lpc2900.cfg:    set _CPUTAPID 0x0596802B
./scripts/target/atheros_ar9331.cfg:    set _CPUTAPID 0x00000001
./scripts/target/stm32l4x.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/stm32l4x.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/at91rm9200.cfg:   set _CPUTAPID 0x05b0203f
./scripts/target/spear3xx.cfg:  set _CPUTAPID 0x07926041
./scripts/target/em357.cfg:      set _CPUTAPID 0x3ba00477
./scripts/target/em357.cfg:      set _CPUTAPID 0x1ba00477
./scripts/target/stm32f7x.cfg:      set _CPUTAPID 0x5ba00477
./scripts/target/stm32f7x.cfg:      set _CPUTAPID 0x5ba02477
./scripts/target/samsung_s3c2440.cfg:   set _CPUTAPID 0x0032409d
./scripts/target/ti-ar7.cfg:    set _CPUTAPID 0x0000100f
./scripts/target/feroceon.cfg:   set _CPUTAPID 0x20a023d3
./scripts/target/lpc3250.cfg:   set _CPUTAPID 0x17900f0f
./scripts/target/stm32f1x.cfg:      set _CPUTAPID 0x3ba00477
./scripts/target/stm32f1x.cfg:      #set _CPUTAPID 0x1ba01478
./scripts/target/stm32f1x.cfg:      #set _CPUTAPID 0x2ba01477
./scripts/target/imx28.cfg:   set _CPUTAPID 0x079264f3
./scripts/target/at91sam7sx.cfg:   set _CPUTAPID 0x3f0f0f0f
./scripts/target/stm32f2x.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/stm32f2x.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/hilscher_netx10.cfg:   set _CPUTAPID 0x25966021
./scripts/target/dsp568037.cfg:   set _CPUTAPID 0x01f2801d
./scripts/target/stm32l0.cfg:    set _CPUTAPID 0x0bc11477
./scripts/target/hilscher_netx500.cfg:   set _CPUTAPID 0x07926021
./scripts/target/imx27.cfg:   set _CPUTAPID 0x07926121
./scripts/target/kx.cfg:      set _CPUTAPID 0x4ba00477
./scripts/target/kx.cfg:      set _CPUTAPID 0x2ba01477
./scripts/target/samsung_s3c4510.cfg:   set _CPUTAPID 0x1f0f0f0f
./scripts/target/at91sam7x256.cfg:   set _CPUTAPID 0x3f0f0f0f
./scripts/board/digi_connectcore_wi-9c.cfg:   set _CPUTAPID 0x07926031
./scripts/board/at91cap7a-stk-sdram.cfg:   set _CPUTAPID 0x40700f0f
./scripts/board/str910-eval.cfg:   set _CPUTAPID 0x25966041
./scripts/board/mini2440.cfg:   set _CPUTAPID 0x0032409d
./scripts/board/at91eb40a.cfg:   set _CPUTAPID 0x1f0f0f0f
./scripts/board/nds32_xc5.cfg:set _CPUTAPID 0x1000063d
./scripts/board/zy1000.cfg:   set _CPUTAPID 0x1f0f0f0f
./scripts/board/hitex_str9-comstick.cfg:   set _CPUTAPID 0x25966041
```

Reply:
Strange, I have been able to flash the bluepill in the past without modifying any file. Although, the last time I tried it didn't work (don't remember why).

Yes, I believe it is possible to modify a variable on the command line. For some boards that use a J-Link there is a workaround, see:

https://github.com/tinygo-org/tinygo/blob/a9ba6ebad9ad85cacd9674a1af9229489a811f68/compileopts/config.go#L264-L270
```go
	if c.Target.OpenOCDTransport != "" && c.Target.OpenOCDTransport != "swd" {
		return nil, fmt.Errorf("unknown OpenOCD transport: %#v", c.Target.OpenOCDTransport)
	}
	// [...]
	if c.Target.OpenOCDTransport != "" {
		args = append(args, "-c", "transport select "+c.Target.OpenOCDTransport)
	}
```

This inserts a custom command (`transport select swd`). I think you can also set a variable that way. For example, we could add a `"cputapid"` key to the JSON file that if set, will modify that variable (possibly to 0).

Reply:
Right now, I've written into the main.go to regex match errors given from openocd, and rerun the openocd subprocess.  Once ever fifty flashes (or so, this is somewhat random)... openocd wont flash the device at all. I found that this has to do with how the st-link performs a software reset to the device, and I've provided a detection and fix for this also.  Replicating the issue is tough.  I have to keep re-flashing until I get the error, and test the fix.  I think I'm going to do a pull request soon.  Right now, I've written the code in main.go, and if there is another place/file you prefer the code changes, I am totally open for suggestions.

I would like to learn how to add support for the nucleo-f303, but I think that's going to have to be another pull request, until I learn more of the code base.

I'll make the pull request today, so you can see where I'm at.

Reply:
pull request is here : https://github.com/tinygo-org/tinygo/pull/1217

Reply:
This sounds like a bug in OpenOCD. And because it's not so frequent (and there is a workaround), I'm not sure whether we should be working around it in such a way. Generally trying to recognize errors from an error log with a regex seems way too brittle (it can break with an update, for example).

Reply:
The error of the CPUTAPID could be fixed by just prepending -c "set CPUTAPID 0" before the board cfg (awesome, easy)  Since tinygo makes a subprocess call to openocd, I would argue that tinygo is responsible for errors that occur while making these calls. I agree, regex is not the best solution, and a better solution would be if openocd had specific error codes that we could just cross check on an exit against some error code table (that might exist, I'll check, but I doubt it).

The other error from the software reset pin/button is not an openocd error. It's an error with the st-link being in able to reset connected device (via software).  As I was trying to make a simple fix for the CPUTAPID, I had gotten this new problem that kept me from just writing tingo code several times.  I think that is unfortunate.  I've just been wanting to simply write code, and I've been slammed with trying to streamline.  The error is significant enough that it can stop someone from working with their blue pill for long periods with this error that becomes consistent, and maybe abandon the blupill all together with tinygo if it becomes frustrating enough.  If there is a regex performance hit from Tiny-Go it's only going to be when flashing with openocd, I think it's relatively small, and worth it under the circumstances.

I think there might be a better solution too.  I'm just not sure what it is.  I at least wanted to provide something that works for now, and if you're going to call openocd as a subprocess you're always going to run the risk of a break in the future, why not make it much easier for newcomers?

I like the idea of adding to the bluepill.json, and maybe changing the args for openocd. I just bought a brand new official $30 st-link and I can't even get the thing to flash at all right now... which is a bummer because I was hoping that might explain why the CPUTAPID has a specific address that openocd sets.  I'll dive a little deeper and see what I can find if I can even get the new st-link to work (I bought the damn thing from ST and I am having nothing but problems).

Another option might be to add an additional flag to Tiny-Go that will allow for higher level error handling?  -ocd-error-handling .  My goal is to flash raw stm33f103's, so having an error handler in the language would be nice if that is something that could be considered. 

Thank you guys for everything so far.

Reply:
@aykevl What device are you using to flash the bluepill? Are you using the "official" st-link, the more expensive j-link, or the cheaper st-link?

Reply:
I have used one of the cheap ST-Link v2 clones you can find on AliExpress. I believe they have the official firmware.

Reply:
@aykevl thanks again, closing here, and will submit another pull request once I've had a chance to edit the json and fiddle with the bluepill json configuration and openOCDConfig constructor.